admiral/pkg/clusters/serviceentry.go

```diff
@@ -770,44 +772,101 @@ func modifyServiceEntryForNewServiceOrPod(
-			// endpoints.
-			destinations, err := getAllVSRouteDestinationsByCluster(
+
+			sourceClusterLocality, err = getClusterRegion(remoteRegistry, sourceCluster, rc)
+			if err != nil {
+				ctxLogger.Warnf(common.CtxLogFormat, "VSBasedRouting",
+					deploymentOrRolloutName, eventNamespace, sourceCluster, err)
+			}
+
+			ingressDestinations, err = getAllVSRouteDestinationsByCluster(
 				ctxLogger,
 				serviceInstance,
 				meshDeployAndRolloutPorts,
 				sourceWeightedServices[sourceCluster],
 				sourceRollouts[sourceCluster],
-				sourceDeployments[sourceCluster],
-				remoteRegistry,
-				sourceIdentity,
-				env)
+				sourceDeployments[sourceCluster])
+
 			if err != nil {
 				ctxLogger.Errorf(common.CtxLogFormat, "getAllVSRouteDestinationsByCluster",
 					deploymentOrRolloutName, eventNamespace, sourceCluster, err)
 				modifySEerr = common.AppendError(modifySEerr, err)
-			} else if len(destinations) == 0 {
+			} else if len(ingressDestinations) == 0 {
 				ctxLogger.Warnf(common.CtxLogFormat, "getAllVSRouteDestinationsByCluster",
 					deploymentOrRolloutName, eventNamespace, sourceCluster,
 					"No RouteDestinations generated for VS based routing ")
 			} else {
-				sourceClusterToDestinations[sourceCluster] = destinations
-				drHost := fmt.Sprintf("*.%s%s", eventNamespace, common.DotLocalDomainSuffix)
-				sourceClusterToDRHosts[sourceCluster] = map[string]string{
-					eventNamespace + common.DotLocalDomainSuffix: drHost,
+				for key, value := range ingressDestinations {
+					inClusterDestinations[key] = value
 				}
 			}
+
+			drHost := fmt.Sprintf("*.%s%s", eventNamespace, common.DotLocalDomainSuffix)
+			sourceClusterToDRHosts[sourceCluster] = map[string]string{
+				eventNamespace + common.DotLocalDomainSuffix: drHost,
+			}
 		} else {
 			ctxLogger.Infof(common.CtxLogFormat, "VSBasedRouting",
 				"", "", sourceCluster,
 				"Discovery phase: VS based routing disabled for cluster")
 		}
+
+		//Discovery Phase: process ingress routing destinations with dns prefixes based on GTP. No GTP weights are updated for ingress destinations
+		if common.DoVSRoutingForCluster(sourceCluster) {
+			err = processGTPAndAddWeightsByCluster(ctxLogger,
+				remoteRegistry,
+				sourceIdentity,
+				env,
+				sourceClusterLocality,
+				ingressDestinations,
+				false)
+			if err != nil {
+				ctxLogger.Errorf(common.CtxLogFormat, "processGTPAndAddWeightsByCluster",
+					deploymentOrRolloutName, eventNamespace, sourceCluster, err)
+				modifySEerr = common.AppendError(modifySEerr, err)
+			}
+			sourceClusterToDestinations[sourceCluster] = ingressDestinations
+		} else {
+			ctxLogger.Infof(common.CtxLogFormat, "processGTPAndAddWeightsByCluster",
+				"", "", sourceCluster,
+				"Discovery phase: Skipped GTP processing as VS based routing disabled for cluster")
+		}
+
+		//Discovery Phase: process in-cluster routing destinations with dns prefixes and weights based on GTP
+		if common.DoVSRoutingInClusterForCluster(sourceCluster) {
```

Comment:
@vrushalijoshi , can you please add this check as well here? `common.DoVSRoutingInClusterForIdentity(sourceIdentity)`